repair: Introduce Host and DC filter support

Deexie opened the pull request:
Currently, the tablet repair scheduler repairs all replicas of a tablet. It does not support hosts or DCs selection. It should be enough for most cases. However, users might still want to limit the repair to certain hosts or DCs in production. https://github.com/scylladb/scylladb/pull/21985 added the preparation work to add the config options for the selection. This patch adds the hosts or DCs selection support.

Fixes https://github.com/scylladb/scylladb/issues/22417

New feature. No backport is needed.

Deexie commented:
Based on https://github.com/scylladb/scylladb/pull/22508. I didn't amend the original commit, but added new on top. To be squashed

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14880/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14880/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_custom/test_tablet_repair_scheduler
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14241/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6229/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3109/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14880/)
#### Failed Tests (24/41027):
- [test_twcs_multiple_sstables_during_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/compaction_additional_test/TestTimeWindowDataSegregation/Tests___dtest___test_twcs_multiple_sstables_during_bootstrap) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_twcs_multiple_sstables_during_bootstrap)
- [test_data_distribution_balance[3-IncrementalCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest___test_data_distribution_balance_3_IncrementalCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance[3-IncrementalCompactionStrategy])
- [test_twcs_multiple_sstables_during_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/compaction_additional_test/TestTimeWindowDataSegregation/Tests___dtest___test_twcs_multiple_sstables_during_bootstrap_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_twcs_multiple_sstables_during_bootstrap)
- [test_replay_after_schema_change](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/batch_test/TestBatch/Tests___dtest_with_tablets___test_replay_after_schema_change) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replay_after_schema_change)
- [test_cast_varint](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/cql_cast_test/TestCQLCast/Tests___dtest_with_tablets___test_cast_varint) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cast_varint)
- [test_data_distribution_balance[4-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest_with_tablets___test_data_distribution_balance_4_LeveledCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance[4-LeveledCompactionStrategy])
- [test_data_distribution_balance[3-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest_with_tablets___test_data_distribution_balance_3_TimeWindowCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance[3-TimeWindowCompactionStrategy])
- [test_data_distribution_balance[3-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest_with_tablets___test_data_distribution_balance_3_TimeWindowCompactionStrategy__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance[3-TimeWindowCompactionStrategy])
- [test_data_distribution_balance[4-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest_with_tablets___test_data_distribution_balance_4_LeveledCompactionStrategy__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance[4-LeveledCompactionStrategy])
- [test_data_distribution_balance[4-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest___test_data_distribution_balance_4_TimeWindowCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance[4-TimeWindowCompactionStrategy])
#### Build Details:
- Duration: 7 hr 8 min
- Builder: spider4.cloudius-systems.com

Deexie commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14880/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14880/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/tablets_test 🔹 topology_custom/test_tablet_repair_scheduler ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14241/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6229/) ❌ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3109/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14880/)
> #### Failed Tests (24/41027):
> 
>     * [test_twcs_multiple_sstables_during_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/compaction_additional_test/TestTimeWindowDataSegregation/Tests___dtest___test_twcs_multiple_sstables_during_bootstrap) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_twcs_multiple_sstables_during_bootstrap)
> 
>     * [test_data_distribution_balance[3-IncrementalCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest___test_data_distribution_balance_3_IncrementalCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance%5B3-IncrementalCompactionStrategy%5D)
> 
>     * [test_twcs_multiple_sstables_during_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/compaction_additional_test/TestTimeWindowDataSegregation/Tests___dtest___test_twcs_multiple_sstables_during_bootstrap_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_twcs_multiple_sstables_during_bootstrap)
> 
>     * [test_replay_after_schema_change](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/batch_test/TestBatch/Tests___dtest_with_tablets___test_replay_after_schema_change) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replay_after_schema_change)
> 
>     * [test_cast_varint](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/cql_cast_test/TestCQLCast/Tests___dtest_with_tablets___test_cast_varint) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cast_varint)
> 
>     * [test_data_distribution_balance[4-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest_with_tablets___test_data_distribution_balance_4_LeveledCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance%5B4-LeveledCompactionStrategy%5D)
> 
>     * [test_data_distribution_balance[3-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest_with_tablets___test_data_distribution_balance_3_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance%5B3-TimeWindowCompactionStrategy%5D)
> 
>     * [test_data_distribution_balance[3-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest_with_tablets___test_data_distribution_balance_3_TimeWindowCompactionStrategy__2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance%5B3-TimeWindowCompactionStrategy%5D)
> 
>     * [test_data_distribution_balance[4-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest_with_tablets___test_data_distribution_balance_4_LeveledCompactionStrategy__2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance%5B4-LeveledCompactionStrategy%5D)
> 
>     * [test_data_distribution_balance[4-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14880/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest___test_data_distribution_balance_4_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance%5B4-TimeWindowCompactionStrategy%5D)
> 
> 
> #### Build Details:
> 
>     * Duration: 7 hr 8 min
> 
>     * Builder: spider4.cloudius-systems.com

https://github.com/scylladb/scylladb/issues/22643
https://github.com/scylladb/scylladb/issues/22645
https://github.com/scylladb/scylladb/issues/22646
https://github.com/scylladb/scylladb/issues/22647

scylladb-promoter commented:
#### :red_circle: CI State: UNSTABLE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14904/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14904/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_custom/test_tablet_repair_scheduler
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6250/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14261/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3129/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14904/)
#### Build Details:
- Duration: 10 hr
- Builder: spider5.cloudius-systems.com

asias commented on service/topology_coordinator.cc:
```diff
@@ -1464,11 +1464,11 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                                 }
                                 auto dc = topo.get_datacenter(replica.host);
-                                if (!dcs_filter.contains(dc)) {
+                                if (!dcs_filter.empty() && !dcs_filter.contains(dc)) {
```
Did you add test to test the case hosts != empty and dcs == empty?

asias commented on test/boost/tablets_test.cc:
```diff
@@ -132,7 +132,7 @@ SEASTAR_TEST_CASE(test_tablet_metadata_persistence) {
                     },
                     db_clock::now(),
-                    locator::tablet_task_info::make_auto_repair_request(),
+                    locator::tablet_task_info::make_auto_repair_request("", "dc1,dc2"),
```
Could you explain why  the default no dc host selection is not enough?

Deexie replied:
It is enough. When adding previous changes to system.tablets (e.g. resize_task_info) I was told to add it to this test to make sure it will be persisted. I thought it's worth doing here, too.

asias commented on test/topology_custom/test_tablet_repair_scheduler.py:
```diff
@@ -190,3 +193,87 @@ async def del_repair_task():
+
+@pytest.mark.asyncio
+@skip_mode('release', 'error injections are not supported in release mode')
+async def test_tablet_repair_hosts_filter(manager: ManagerClient):
```
Let's test both hosts and dcs are on.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15001/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15001/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_custom/test_tablet_repair_scheduler
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6328/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3208/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14343/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15001/)
#### Build Details:
- Duration: 8 hr 56 min
- Builder: i-0a63e18fe20adbcc7 (m5d.12xlarge)

denesb commented:
@asias please re-review.

denesb commented on api/storage_service.cc:
```diff
@@ -1697,8 +1697,19 @@ rest_repair_tablet(http_context& ctx, sharded<service::storage_service>& ss, std
 
-        auto res = co_await ss.local().add_repair_tablet_request(table_id, tokens_variant);
+        std::unordered_set<sstring> hosts_vec;
+        boost::split(hosts_vec, hosts, boost::algorithm::is_any_of(","));
```
We try to move away from `boost` ranges, isn't there a way to do this with `std::ranges`?

denesb commented on service/storage_service.cc:
```diff
@@ -6277,13 +6277,15 @@ future<bool> storage_service::exec_tablet_update(service::group0_guard guard, st
-future<std::unordered_map<sstring, sstring>> storage_service::add_repair_tablet_request(table_id table, std::variant<utils::chunked_vector<dht::token>, all_tokens_tag> tokens_variant) {
+future<std::unordered_map<sstring, sstring>> storage_service::add_repair_tablet_request(table_id table,
+        std::variant<utils::chunked_vector<dht::token>, all_tokens_tag> tokens_variant,
+        sstring hosts_filter, sstring dcs_filter) {
```
Why not make these `std::unordered_set<hos_id>` and `std::unordered_set<sstring>` respectively? This way the interface already guarantees de-duplication and that host-ids are parseable.

Deexie replied:
This way we would need to serialize it a moment later. It would probably be better from the interface perspective, but now we do not need to serialize the lists at all

denesb replied:
I think the interface clarity is worth it. Starting a repair is not on the hot path.

denesb commented on api/storage_service.cc:
```diff
@@ -1701,12 +1701,14 @@ rest_repair_tablet(http_context& ctx, sharded<service::storage_service>& ss, std
         auto dcs = req->get_query_param("dcs_filter");
```
Please do the TODO :)

Deexie commented:
- squash commits
- use `std::ranges::view::split` instead of `boost::split`
- add checks whether required node were actually repaired
- use `std::unordered_set` for hosts and dcs lists
   - I did it also for tablet_task_info, assuming we want to have proper types whenever possible, but I'm uncertain if it was a good decision, because its ser/deserialization would happen on each version change

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15073/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15073/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_custom/test_tablet_repair_scheduler
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3275/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14409/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6392/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15073/)
#### Build Details:
- Duration: 12 hr
- Builder: i-09d4482debbca174e (m5ad.12xlarge)

asias commented on locator/tablets.cc:
```diff
@@ -26,6 +26,8 @@
 #include <seastar/coroutine/maybe_yield.hh>
 #include <type_traits>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/algorithm/string/classification.hpp>
```
Do we still need the boost ones?

asias commented:
> * squash commits
> * use `std::ranges::view::split` instead of `boost::split`
> * add checks whether required node were actually repaired
> * use `std::unordered_set` for hosts and dcs lists
>   
>   * I did it also for tablet_task_info, assuming we want to have proper types whenever possible, but I'm uncertain if it was a good decision, because its ser/deserialization would happen on each version change

I think this is ok. Most of the cases, the dcs and hosts filter will be empty. 

The new version looks good.

Deexie commented:
- drop boost includes

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15193/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15193/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_custom/test_tablet_repair_scheduler
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14507/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6486/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3371/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15193/)
#### Failed Tests (6/41537):
- [cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.1)
- [cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.2)
- [cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.3)
- [cqlpy.cassandra_tests/validation/entities/uf_types_test.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.release.1)
- [cqlpy.cassandra_tests/validation/entities/uf_types_test.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.release.2)
- [cqlpy.cassandra_tests/validation/entities/uf_types_test.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.release.3)
#### Build Details:
- Duration: 12 hr
- Builder: i-0e6c5d16139a58fd6 (m5ad.12xlarge)

denesb commented:
> * [cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_dev_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.1)
> 
> * [cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_dev_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.2)
> 
> * [cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.3)
> 
> * [cqlpy.cassandra_tests/validation/entities/uf_types_test.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.release.1)
> 
> * [cqlpy.cassandra_tests/validation/entities/uf_types_test.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_release_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.release.2)
> 
> * [cqlpy.cassandra_tests/validation/entities/uf_types_test.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_release_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.release.3)

https://github.com/scylladb/scylladb/issues/22799

scylladb-promoter commented:
#### :red_circle: CI State: UNSTABLE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15281/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15281/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_custom/test_tablet_repair_scheduler
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6535/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14554/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3418/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15281/)
#### Build Details:
- Duration: 10 hr
- Builder: spider5.cloudius-systems.com

denesb commented:
@Deexie please rebase, there are conflicts.

Deexie commented:
- rebase

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15299/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15299/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_custom/test_tablet_repair_scheduler
#### Failed Tests (22/281):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15299/testReport/junit/boost.tablets_test.test_load_balancing_merge_colocation_with_random_load.5.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [test_tablet_manual_repair_async.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15299/testReport/junit/topology_custom/test_tablet_repair_scheduler/test_tablet_manual_repair_async_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_manual_repair_async.debug.3)
- [test_tablet_manual_repair_async.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15299/testReport/junit/topology_custom/test_tablet_repair_scheduler/test_tablet_manual_repair_async_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_manual_repair_async.debug.4)
- [test_tablet_manual_repair_async.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15299/testReport/junit/topology_custom/test_tablet_repair_scheduler/test_tablet_manual_repair_async_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_manual_repair_async.dev.1)
- [test_tablet_manual_repair_async.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15299/testReport/junit/topology_custom/test_tablet_repair_scheduler/test_tablet_manual_repair_async_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_manual_repair_async.dev.2)
- [test_tablet_manual_repair_async.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15299/testReport/junit/topology_custom/test_tablet_repair_scheduler/test_tablet_manual_repair_async_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_manual_repair_async.dev.3)
- [test_tablet_manual_repair_async.dev.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15299/testReport/junit/topology_custom/test_tablet_repair_scheduler/test_tablet_manual_repair_async_dev_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_manual_repair_async.dev.4)
- [test_tablet_manual_repair_async.dev.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15299/testReport/junit/topology_custom/test_tablet_repair_scheduler/test_tablet_manual_repair_async_dev_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_manual_repair_async.dev.5)
- [test_tablet_manual_repair_async.dev.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15299/testReport/junit/topology_custom/test_tablet_repair_scheduler/test_tablet_manual_repair_async_dev_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_manual_repair_async.dev.6)
- [test_tablet_manual_repair_async.dev.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15299/testReport/junit/topology_custom/test_tablet_repair_scheduler/test_tablet_manual_repair_async_dev_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_manual_repair_async.dev.7)
#### Build Details:
- Duration: 2 hr 1 min
- Builder: i-0e572a854fa6edf3b (m5ad.12xlarge)

Deexie commented:
- fix bad rebase in tests

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15318/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15318/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_custom/test_tablet_repair_scheduler
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3448/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14586/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6566/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15318/)
#### Build Details:
- Duration: 6 hr 12 min
- Builder: spider4.cloudius-systems.com

asias commented:
Since 2025.1 is branched, we need this for 2025.1. Removing the backport/none label.

tgrabiec commented on service/topology_coordinator.cc:
```diff
@@ -1455,9 +1455,32 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                           }
+                        }
+                        if (!dst_opt) {
+                            throw std::runtime_error(fmt::format("Could not find host to perform repair tablet={} replica={} hosts_filter={} dcs_filter={}",
```
There's no guarantee that every tablet will have at least 1 replica which matches the filter. Is anything preventing this failure? 

Failing here will block the state machine until manual intervention.

asias replied:
Even if we do the check before submitting the request, it is possible when the locator::tablet_transition_stage::repair runs, there is no valid replicas after the filter. 

I think what we can do is:
1) if the filter is enabled:
Skip the row level repair and move to the next stage in case there is less than 2 replicas after the filter. Ensure the repair time is not updated (we will not update with current code in this case already). 

2) if the filter is not enabled
Throw as what we do now. This is not supposed to happen. 

@Deexie

Deexie replied:
I do not understand how it is supposed to block the state machine. It runs under `advance_in_background`, won't `tablet_state.repair` just hold the failed future that will be handled here:
https://github.com/Deexie/scylla/blob/5545289bfae0508fd72b655ddd2fef6c5dbe60a2/service/topology_coordinator.cc#L1441-L1449
?

Deexie replied:
Anyway, currently such a repair will be retried forever, so I guess this should be replaced with `co_return`, no?

tgrabiec replied:
> I do not understand how it is supposed to block the state machine. It runs under `advance_in_background`, won't `tablet_state.repair` just hold the failed future that will be handled here: https://github.com/Deexie/scylla/blob/5545289bfae0508fd72b655ddd2fef6c5dbe60a2/service/topology_coordinator.cc#L1441-L1449 ?

Right, it will just make the repair will be retried forever.

tgrabiec replied:
> Anyway, currently such a repair will be retried forever, so I guess this should be replaced with `co_return`, no?

I think so, yes.

I can see the logic is duplicated here and in the repair handler. Would be nice to abstract it and share it to avoid discrepancies.

E.g: 

```
for (auto& replica : tinfo.replicas) {
   if (selected_by_filter(replica, tinfo.repair_task_info, topo)) {

```

and:

```
auto dst_opt = get_primary(replica, tinfo.repair_task_info, topo);
```

...which combines the filter.

Also, in case we have a DC filter, we lost the round-robin selection of primary replica, we will always select the first one. We should probably bring it back.

Deexie replied:
https://github.com/scylladb/scylladb/pull/23101
